platform/stm32f4xx/export/platform/i2c_bus.hpp

```diff
@@ -0,0 +1,648 @@
+    static constexpr auto pick_i2c();
+    static constexpr auto pick_ev_irqn();
+    static constexpr auto pick_er_irqn();
+
+    //! Communication direction
+    // TODO Add slave support
+    enum direction {
+        MASTER_TX,
+        MASTER_RX,
+    } m_direction;
+
+    // Device status flags
+
+    //! Bit set if device initialized.
+    static constexpr uint8_t m_inited     = 0x1;
+
+    // Device status methods
+    inline bool inited() const     { return (m_status & m_inited)  != 0; }
+
+    inline void set_inited()       { m_status |= m_inited; }
+    inline void clear_inited()     { m_status &= ~(m_inited); }
+
+    //! Setup xfer in IRQ mode
+    ecl::err i2c_setup_xfer_irq();
+
+    //! Operations for POLL mode
+    ecl::err i2c_transmit_poll();
+    ecl::err i2c_receive_poll();
+
+    //! Handles IRQ events from a bus.
+    void irq_ev_handler();
+    //! Handles IRQ events (error) from a bus.
+    void irq_er_handler();
+
+    handler_fn      m_event_handler; //! Handler passed via set_handler().
+    const uint8_t   *m_tx;           //! Transmit buffer.
+    size_t          m_tx_size;       //! TX buffer size.
+    size_t          m_tx_left;       //! Left to send in TX buffer.
+    uint8_t         *m_rx;           //! Receive buffer.
+    size_t          m_rx_size;       //! RX buffer size.
+    size_t          m_rx_left;       //! Left to receive in RX buffer.
+    uint8_t         m_status;        //! Tracks device status.
+
+    uint16_t 		m_own_addr;		 //! device address in slave mode
+    uint16_t 		m_slave_addr;	 //! address of slave to communicate
+};
+
+template< class i2c_config >
+i2c_bus< i2c_config >::i2c_bus()
+	:m_event_handler{}
+	,m_tx{nullptr}
+	,m_tx_size{0}
+	,m_tx_left{0}
+	,m_rx{nullptr}
+	,m_rx_size{0}
+	,m_rx_left{0}
+	,m_status{0}
+	,m_slave_addr{0}
+{
+
+}
+
+template< class i2c_config >
+i2c_bus< i2c_config >::~i2c_bus()
+{
+
+}
+
+template < class i2c_config >
+ecl::err i2c_bus <i2c_config>::init()
+{
+    if (inited()) {
+        return ecl::err::ok;
+    }
+
+    constexpr auto i2c = pick_i2c();
+    constexpr auto rcc_periph = pick_rcc();
+    constexpr auto rcc_fn = pick_rcc_fn();
+
+    rcc_fn(rcc_periph, ENABLE);
+
+    constexpr auto init_const_obj = i2c_config::m_init_obj;
+    auto init_obj = init_const_obj;
+    I2C_Init(i2c, &init_obj);
+
+    I2C_Cmd(i2c, ENABLE);
+
+    if (i2c_config::m_mode == i2c_mode::IRQ) {
+        constexpr auto irqn_ev = pick_ev_irqn();
+        constexpr auto irqn_er = pick_er_irqn();
+
+        auto lambda_ev = [this]() {
+            this->irq_ev_handler();
+        };
+
+        auto lambda_er = [this]() {
+            this->irq_er_handler();
+        };
+
+        IRQ_manager::subscribe(irqn_ev, lambda_ev);
+        IRQ_manager::subscribe(irqn_er, lambda_er);
+    }
+
+    set_inited();
+
+    return ecl::err::ok;
+}
+
+template< class i2c_config >
+void i2c_bus< i2c_config >::set_rx(uint8_t *rx, size_t size)
+{
+    ecl_assert(inited());
```

Comment:
Using a regular assert in the platform introduces a cyclic dependency. The platform module depends on the assert module, which is in turn depends on iostream. Iostream depends on a console driver which is a part of the platform.
 
Also, regular assert cannot be used in irq routines and it doesn't guarantee that console subsystem is initialized prior to its call.

Separate platform-level assert must be introduced. For now, you can just exit from a function, or set error bit.